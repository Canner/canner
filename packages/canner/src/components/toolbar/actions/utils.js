import {
  forEach,
} from 'lodash';

export function flattenItems(items) {
  const fields = [];

  function loop(item, keyName = '') {
    let newKeyName = keyName;
    if (item.keyName) {
      newKeyName = `${newKeyName}${newKeyName ? '.' : ''}${item.keyName}`;
    }

    if (item.items && !item.items.type) {
      forEach(item.items, (item) => {
        loop(item, newKeyName);
      });
    } else if (item.items && item.items.type && item.items.items) {
      forEach(item.items.items, (item) => {
        loop(item, newKeyName);
      });
    } else {
      fields.push({
        ...item,
        keyName: newKeyName,
      });
    }
  }

  forEach(items, (item) => {
    loop(item, '');
  });

  return fields;
}
