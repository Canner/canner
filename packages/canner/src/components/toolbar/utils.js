// @flow

export function findAlwaysDisplayedFilterIndexes(filters: Array<Object>): Array<number> {
  const result = [];
  filters.forEach((filter, index) => {
    if (filter.alwaysDisplay) {
      result.push(index);
    }
  });
  return result;
}