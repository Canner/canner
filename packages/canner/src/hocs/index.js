import errorCatch from './errorCatch';
import graphqlQuery from './graphqlQuery';
import withCanner from './withCanner';
import withCannerLayout from './withCannerLayout';
import withCache from './withCache';
export default {
  errorCatch,
  graphqlQuery,
  withCanner,
  withCannerLayout,
  withCache
};