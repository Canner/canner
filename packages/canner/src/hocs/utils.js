// @flow
import {mapValues, get, isPlainObject, isArray} from 'lodash';

export function getValue(value: any, idPathArr: Array<string>) {
  return idPathArr.reduce((result: any, key: string) => {
    if (isPlainObject(result)) {
      if ('edges' in result && 'pageInfo' in result) {
        return get(result, ['edges', key, 'node']);
      }
      return get(result, key);
    } else if (isArray(result)) {
      return get(result, key);
    } else {
      return result;
    }
  }, value);
}

export function parseConnectionToNormal(value: any) {
  if (isPlainObject(value)) {
    if (value.edges && value.pageInfo) {
      return value.edges.map(edge => parseConnectionToNormal(edge.node));
    }
    return mapValues(value, item => parseConnectionToNormal(item));
  } else if (isArray(value)) {
    return value.map(item => parseConnectionToNormal(item))
  } else {
    return value;
  }
}

export function defaultValue(type: string, relation: any) {
  switch (type) {
    case 'connection': {
      return {
        edges: [],
        pageInfo: {
          hasNextPage: false,
          hasPreviousPage: false
        }
      }
    }
    case 'array': {
      return [];
    }
    case 'object': {
      return {};
    }
    case 'boolean': {
      return false;
    }
    case 'number': {
      return 0;
    }
    case 'string': {
      return '';
    }
    case 'relation': {
      if (relation.type === 'toMany') {
        return {
          edges: [],
          pageInfo: {
            hasNextPage: false,
            hasPreviousPage: false
          }
        };
      } else {
        return null;
      }
    }
    case 'image':
    case 'file': {
      return {
        url: '',
        contentType: '',
        name: '',
        size: 0,
        __typename: null
      }
    }
    case 'geoPoint': {
      return {
        placeId: '',
        address: '',
        lat: 122,
        lng: 23
      };
    }
    default: {
      return null;
    }
  }
}

export function paginate(rootValue: Object, field: string, currentPage: number, pageSize: number = 10) {
  const copyValue = JSON.parse(JSON.stringify(rootValue));
  const index = (currentPage - 1) * pageSize;
  copyValue[field].edges = copyValue[field].edges.slice(index, index + pageSize);
  return copyValue;
}

export function filterByWhere(rootValue: Object, field: string, where: Object) {
  const copyValue = JSON.parse(JSON.stringify(rootValue));
  copyValue[field].edges = copyValue[field].edges.filter(edge => {
    return Object.keys(where).reduce((result, key) => {
      const condition = where[key];
      const data = edge.node[key];
      return result && passCondition(data , condition);
    }, true);
  });
  return copyValue;
}

export function passCondition(data: any, condition: Object) {
  return Object.keys(condition).reduce((result, conditionKey) => {
    let isPass = true;
    const conditionValue = condition[conditionKey];
    switch(conditionKey) {
      case 'contains':
        isPass = data.indexOf(conditionValue) !== -1;
        break;
      case 'eq':
        isPass = data === conditionValue;
        break;
      case 'gt':
        isPass = data > conditionValue;
        break;
      case 'gte':
        isPass = data >= conditionValue;
        break;
      case 'lt':
        isPass = data < conditionValue;
        break;
      case 'lte':
        isPass = data <= conditionValue;
        break;
    }
    return result && isPass;
  }, true);
}