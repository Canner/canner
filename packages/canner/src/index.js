export default from './components';
export {hocs} from './hocs';
export {default as Provder} from './components/Provider';
export {default as ReactRouterProvider} from './components/routeProvider/ReactRouterProvider';
