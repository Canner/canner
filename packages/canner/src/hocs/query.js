// @flow

import * as React from 'react';
import {List} from 'react-content-loader';
import Toolbar from '../components/toolbar/index';
import {Icon, Spin} from 'antd';
import {mapValues, isNil} from 'lodash';
import {getValue, defaultValue} from './utils';
import {withApollo} from 'react-apollo';
import type {HOCProps} from './types';
type State = {
  value: any,
  rootValue: any,
  originRootValue: any,
  isFetching: boolean,
}

@withApollo
export default function withQuery(Com: React.ComponentType<*>) {
  // this hoc will fetch data;
  return class ComponentWithQuery extends React.PureComponent<HOCProps, State> {
    key: string;
    subscription: any;

    constructor(props: HOCProps) {
      super(props);
      this.state = {
        value: null,
        rootValue: null,
        originRootValue: null,
        isFetching: true
      };
      this.key = props.refId.getPathArr()[0];
    }

    componentDidMount() {
      this.queryData().then(() => {
        this.subscribe();
      });
    }

    componentWillUnmount() {
      this.unsubscribe();
    }

    componentDidUpdate(prevProps: HOCProps) {
      const {refId} = prevProps;
      if (refId.toString() !== this.props.refId.toString()) {
        // refetch when route change
        this.unsubscribe();
        this.queryData(this.props).then(() => {
          this.subscribe();
        });
      }
    }

    unsubscribe = () => {
      if (this.subscription) {
        this.subscription.unsubscribe();
      }
    }

    subscribe = () => {
      const {subscribe} = this.props;
      const subscription = subscribe(this.key, result => {
        const {data, rootValue} = result;
        this.updateData(data, rootValue);
      });
      this.subscription = subscription;
    }

    queryData = (props?: HOCProps): Promise<*> => {
      const {fetch} = props || this.props;
      this.setState({
        isFetching: true
      });
      return fetch(this.key).then(result => {
        const {data, rootValue} = result;
        this.updateData(data, rootValue);
      });
    }

    updateData = (data: any, rootValue: any) => {
      const {refId} = this.props;
      const newValue = getValue(data, refId.getPathArr());
      this.setState({
        originRootValue: data,
        rootValue,
        value: newValue,
        isFetching: false
      });
    }

    updateQuery = (paths: Array<string>, args: Object) => {
      const {updateQuery} = this.props;
      updateQuery(paths, args)
        .then(rewatch => {
          if (rewatch) {
            this.unsubscribe();
            this.queryData();
            this.subscribe();
          } else {
            this.queryData();
          }
        });
    }

    render() {
      const {value, isFetching, rootValue, originRootValue} = this.state;
      const {toolbar, query, refId, items, type, path, relation, pattern, keyName, request, deploy, nullable} = this.props;
      if (!originRootValue) {
        return <List style={{maxWidth: '600px'}}/>;
      }
      if (pattern === 'array') {
        const queries = query.getQueries(path.split('/')).args || {pagination: {first: 10}};
        const variables = query.getVairables();
        const args = mapValues(queries, v => variables[v.substr(1)]);

        return (
          <Toolbar items={items}
            toolbar={toolbar}
            args={args}
            query={query}
            refId={refId}
            keyName={keyName}
            originRootValue={originRootValue}
            rootValue={rootValue}
            getValue={getValue}
            updateQuery={this.updateQuery}
            request={request}
            deploy={deploy}
          >
            <SpinWrapper isFetching={isFetching}>
              <Com {...this.props} />
            </SpinWrapper>
          </Toolbar>
        );
      } else if (type === 'relation' && relation.type === 'toOne') {
        return (
            <Com {...this.props} showPagination={true} rootValue={rootValue} value={(value && value.id) ? value : defaultValue(type, relation)} />
        );
      } else if (type === 'relation' && relation.type === 'toMany') {
        return (
            <Com {...this.props} showPagination={true} rootValue={rootValue} value={isNil(value) ? defaultValue('array', undefined, nullable) : value}/>
        );
      }

      return (
        <Com {...this.props} showPagination={true} rootValue={rootValue} value={isNil(value) ? defaultValue(type, relation, nullable) : value} />
      );
    }
  };
}

const antIcon = <Icon type="loading" style={{fontSize: 24}} spin />;

function SpinWrapper({
  isFetching,
  children,
  ...props
}: {
  isFetching: boolean,
  children: React.Element<*>
}): React.Element<*> {
  return (
    <Spin indicator={antIcon} spinning={isFetching}>
      {React.cloneElement(children, props)}
    </Spin>
  )
}

export function shouldUpdate(value: any, newValue: any) {
  return value != newValue;
}
