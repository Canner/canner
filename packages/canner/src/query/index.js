// @flow

import {schemaToQueriesObject, objectToQueries} from './utils';
import {get, set, mapKeys} from 'lodash';

export class Query {
  schema: Object
  queries: Object
  variables: Object

  constructor({schema}: {schema: Object}) {
    this.schema = schema;
    const {queriesObj, variables} = schemaToQueriesObject(schema);
    this.variables = variables;
    this.queries = queriesObj;
  }

  updateQueries = (pathArr: Array<string>, field: string, value: any) => {
    const path = `${pathArr.join('.fields.')}.${field}`;
    if (field === 'args') {
      const args = get(this.queries, path);
      this.variables = {...this.variables, ...Object.keys(args).reduce((result: Object, key: string) => {
        result[args[key]] = value[key];

        return result;
      }, {})};
    } else {
      set(this.queries, path, value);
    }
  }

  getQueries = (pathArr: Array<string>): Object => {
    if (!pathArr || pathArr.length === 0) {
      return this.queries;
    }
    const path = pathArr.join('.fields.');
    return get(this.queries, path);
  }

  toGQL = (key?: string): string => {
    const variables = this.getVairables();
    if (key) {
      const obj = this.getQueries([key]);
      return objectToQueries({[key]: obj}, !obj.declareArgs, variables);  
    } else {
      return objectToQueries(this.queries, false, variables);
    }
  }

  getQueryKey = (key: string): string => {
    // return the string which includes the information of this query excluding the fields and declareArgs
    // e.g: postsConnection(first: $postsFirst,where: $postsWhere)
    const queries = this.getQueries([key]);
    delete queries.fields;
    delete queries.declareArgs;
    return objectToQueries({[key]: queries}, false)
  }

  getVairables = () => {
    return mapKeys(this.variables, (value, key) => key.substr(1));
  }
}