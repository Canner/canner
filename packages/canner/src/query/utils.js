// @flow

import pluralize from 'pluralize';
import lowerFirst from 'lodash/lowerFirst';
import upperFirst from 'lodash/upperFirst';
import {merge, mapValues, set} from 'lodash';
import {createSchema} from './schema/utils';
import {types} from './schema/types';

const DEFAULT_FIRST = 10;
const MAX = 999999999;

export function fieldToQueriesObject(field: any): any {
  let queriesObj = {};
  const variables = {};
  let variableTypes = {};
  const type = field.getType();
  switch (type) {
    case types.OBJECT: {
      field.forEach(childField => {
        const qlo = fieldToQueriesObject(childField);
        set(queriesObj, ['fields', childField.getKey()], qlo.queriesObj);
      });
      break;
    }
    case types.ARRAY: {
      field.forEach(childField => {
        const qlo = fieldToQueriesObject(childField);
        set(queriesObj, ['fields', childField.getKey()], qlo.queriesObj);
        merge(variables, qlo.variables);
        merge(variableTypes, qlo.variableTypes);
      });
      if (field.isEntity) {
        const {args, firstKey, afterKey, lastKey, beforeKey, whereKey, orderByKey} = genQuery();
        queriesObj.declareArgs = {
          ...variableTypes,
          [firstKey]: 'Int',
          [afterKey]: 'String',
          [lastKey]: 'Int',
          [beforeKey]: 'String',
          [whereKey]: `${typeKey(field.getKey())}WhereInput`,
          [orderByKey]: `${typeKey(field.getKey())}OrderByInput`
        };
        queriesObj.args = args;
        queriesObj.isPlural = true;
        queriesObj.connection = true;
        queriesObj.alias = field.getKey();
        set(queriesObj, ['fields', 'id'], null);
        variables[firstKey] = MAX;
        variables[whereKey] = {};
        const toolbar = field.getAttr('toolbar');
        const asyncToolbar = toolbar && toolbar.async;
        const defaultSort = toolbar && toolbar.sort && toolbar.sort.defaultSort;
        const permanentFilter = toolbar && toolbar.filter && toolbar.filter.permanentFilter;
        if (asyncToolbar) {
          variables[firstKey] = DEFAULT_FIRST;
        }
        if (permanentFilter) {
          variables[whereKey] = permanentFilter;
        }
        if (defaultSort) {
          variables[orderByKey] = `${defaultSort}_ASC`
        }
        
      }
      break;
    }

    case types.RELATION: {
      set(queriesObj, ['fields', 'id'], null);
      field.forEach(childField => {
        if (childField.getType() !== types.RELATION) {
          const qlo = fieldToQueriesObject(childField);
          set(queriesObj, ['fields', childField.getKey()], qlo.queriesObj);
          merge(variables, qlo.variables);
          merge(variableTypes, qlo.variableTypes);
        }
      });
      // for now, fetch all toMany data
      // if (field.isToMany()) {
      //   const {args, firstKey, afterKey, lastKey, beforeKey, whereKey, orderByKey} = genQuery();
      //   queriesObj.args = args;
      //   queriesObj.isPlural = true;
      //   variables[firstKey] = defaultFirst;
      //   variableTypes = {
      //     ...variableTypes,
      //     [firstKey]: 'Int',
      //     [afterKey]: 'String',
      //     [lastKey]: 'Int',
      //     [beforeKey]: 'String',
      //     [whereKey]: `${typeKey(field.relationTo())}WhereUniqueInput`,
      //     [orderByKey]: `${typeKey(field.relationTo())}WhereUniqueInput`
      //   }
      // }
      break;
    }

    case types.DATETIME:
    case types.FILE:
    case types.GEOPOINT:
    case types.IMAGE: {
      field.forEach(childField => {
        const qlo = fieldToQueriesObject(childField);
        set(queriesObj, ['fields', childField.getKey()], qlo.queriesObj);
      });
      break;
    }
    case types.BOOLEAN:
    case types.NUMBER:
    case types.INT:
    case types.ID:
    case types.STRING:
    default:
      queriesObj = null;
      break;
  }
  return {
    queriesObj,
    variables,
    variableTypes
  };
}

export function genQuery() {
  const firstKey = randomKey();
  const afterKey = randomKey();
  const lastKey = randomKey();
  const beforeKey = randomKey();
  const whereKey = randomKey();
  const orderByKey = randomKey();
  const args = {
    first: firstKey,
    after: afterKey,
    last: lastKey,
    before: beforeKey,
    where: whereKey,
    orderBy: orderByKey
  };
  return {args, firstKey, afterKey, lastKey, beforeKey, whereKey, orderByKey};
}

export function schemaToQueriesObject(schema: any) {
  const rootFields = createSchema(schema);
  const rootVariables = {};
  let rootVariableTypes = {};
  const queriesObj = mapValues(rootFields, (v, key) => {
    const {variables, queriesObj, variableTypes} = fieldToQueriesObject(v);
    merge(rootVariables, variables);
    rootVariableTypes[key] = variableTypes;
    return queriesObj;
  });

  return {
    queriesObj,
    variables: rootVariables
  }
}

export function objectToQueries(o: Object, close: boolean = true, variables?: Object) {

  const result = Object.keys(o).map(key => {
    let query = `${key}`;
    let element = o[key];
    if (!element) {
      return `${query}`;
    }

    if (element.declareArgs) {
      const originElement = {...element};
      const args = originElement.declareArgs;
      delete originElement.declareArgs;
      query = 'query';
      element = {
        args,
        fields: {
          [key]: originElement
        }
      };
      key = 'query';
    }

    if (element.isPlural) {
      query = pluralize.plural(lowerFirst(query));
    }

    if (element.connection) {
      query = `${query}Connection`;
    }

    if (element.alias) {
      query = `${element.alias}: ${query}`;
    }

    if (element.args) {
      if (key === 'query') {
        const args = genDeclareArgs(element.args, variables);
        query = args ? `${query}(${args})` : `${query}`;
      } else {
        const args = genArgs(element.args, variables);
        query = args ? `${query}(${args})` : `${query}`;
      }
    }

    if (element.fields) {
      let {fields} = element;
      if (element.connection) {
        fields = {
          edges: {
            fields: {
              cursor: null,
              node: {
                fields
              }
            }
          },
          pageInfo: {
            fields: {
              hasNextPage: null,
              hasPreviousPage: null
            }
          }
        }
      }
      fields = objectToQueries(fields, true, variables);
      
      query = `${query}${fields}`;
    }
    return `${query}`;
  }).join(' ');
  return close ? `{${result}}` : result;
}

function genArgs(args, variables) {
  return Object.keys(args)
    .filter(key => {
      if (variables && variables[args[key].substr(1)] === undefined) {
        return false;
      }
      return true;
    })
    .map(key => {
      const argValue = args[key];
      return `${key}: ${argValue}`
    }).join(',');
}

function genDeclareArgs(args, variables) {
  return Object.keys(args)
    .filter(key => {
      if (variables && variables[key.substr(1)] === undefined) {
        return false;
      }
      return true;
    })
    .map(key => {
      const argValue = args[key];
      return `${key}: ${argValue}`
    }).join(',');
}

function randomKey () {
  if (process.env.NODE_ENV === 'test') {
    return `$RANDOM_KEY`;
  }
  return `$KEY${Math.random().toString(36).substr(2, 7)}`
}

function typeKey(key) {
  return upperFirst(pluralize.singular(key));
}
