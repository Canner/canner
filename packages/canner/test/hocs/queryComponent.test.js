import Toolbar, {parseOrder, parsePagination, parseWhere} from '../../src/hocs/components/toolbar';
import {getValue, parseConnectionToNormal, defaultValue} from '../../src/hocs/utils';
import * as React from 'react';
import Enzyme, { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import RefId from 'canner-ref-id';
import {Query} from '../../src/query';
import {mapValues} from 'lodash';

Enzyme.configure({ adapter: new Adapter() });

describe('parse order', () => {
  it('should work', () => {
    const args = {
      orderBy: 'age_DESC'
    }
    const {orderField, orderType} = parseOrder(args.orderBy);
    expect(orderField).toBe('age');
    expect(orderType).toBe('DESC');
  });

  it('should return default', () => {
    const args = {

    };
    const {orderField, orderType} = parseOrder(args.orderBy);
    expect(orderField).toEqual(null);
    expect(orderType).toBe('ASC');
  });
});

describe('parse pagination', () => {
  it('should work', () => {
    const pagination = {
      first: 1,
      after: 3
    };

    expect(parsePagination(pagination)).toEqual(pagination);
  });
});

describe('parse where', () => {
  it('string 1', () => {
    const where = {
      name: 'name'
    };
    expect(parseWhere(where)).toEqual({
      name: {
        eq: 'name'
      }
    });
  });

  it('string 2', () => {
    const where = {
      name_eq: 'name'
    };
    expect(parseWhere(where)).toEqual({
      name: {
        eq: 'name'
      }
    });
  })

  it('number 1', () => {
    const where = {
      number: 1
    };
    expect(parseWhere(where)).toEqual({
      number: {
        eq: 1
      }
    });
  })

  it('number 2', () => {
    const where = {
      number_gt: 1
    };
    expect(parseWhere(where)).toEqual({
      number: {
        gt: 1
      }
    });
  });

  it('object 1', () => {
    const where = {
      author: {
        name: 'name'
      }
    };
    expect(parseWhere(where)).toEqual({
      author: {
        name: {
          eq: 'name'
        }
      }
    });
  })
});

describe('toolbar method', () => {
  let props, query, refId, items, originRootValue, toolbar, updateQuery;

  beforeEach(() => {
    const schema = {
      posts: {
        type: 'array',
        items: {
          type: 'object',
          items: {
            title: {
              type: 'string'
            },
            like: {
              type: 'number'
            },
            share: {
              type: 'number'
            }
          }
        }
      }
    };
    query = new Query({
      schema
    });
    updateQuery = jest.fn();
    refId = new RefId('posts');
    originRootValue = {
      posts: {
        edges: [{
          cursor: "0",
          node: {
            id: "0",
            title: "posts1"
          }
        }],
        pageInfo: {
          hasNextPage: true,
          hasPreviousPage: true,
        }
      }
    };
    items = schema.posts.items.items;
    toolbar = {
      async: true,
      sort: {},
      pagination: {},
      filter: {}
    }
    props = {
      items,
      refId,
      query,
      originRootValue,
      toolbar,
      updateQuery,
      getValue,
      keyName: 'posts',
      parseConnectionToNormal,
      defaultValue,
      args: mapValues(query.getQueries(refId.getPathArr()).args, v => query.getVairables()[v]),
    };
  });

  it('should render', () => {
    const wrapper = shallow(<Toolbar {...props}>
      <div></div>
    </Toolbar>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should call updateQuery', () => {
    const wrapper = shallow(<Toolbar {...props}>
      <div></div>
    </Toolbar>);
    wrapper.instance().changeOrder({
      orderField: 'like',
      orderType: 'DESC'
    });
    expect(updateQuery.mock.calls[0][0]).toEqual(['posts']);
    expect(updateQuery.mock.calls[0][1]).toEqual({
      orderBy: 'like_DESC',
      first: 10,
    });
  });

  it('should call updateQuery', () => {
    const wrapper = shallow(<Toolbar {...props}>
      <div></div>
    </Toolbar>);
    wrapper.instance().changeFilter({
      share: {
        gt: 1
      }
    });
    expect(updateQuery.mock.calls[0][0]).toEqual(['posts']);
    expect(updateQuery.mock.calls[0][1]).toEqual({
      first: 10,
      where: {share_gt: 1}
    });
  });

  it('should call updateQuery', () => {
    const wrapper = shallow(<Toolbar {...props}>
      <div></div>
    </Toolbar>);
    wrapper.instance().nextPage();
    expect(updateQuery.mock.calls[0][0]).toEqual(['posts']);
    expect(updateQuery.mock.calls[0][1]).toEqual({
      orderBy: undefined,
      after: "0",
      first: 10,
      last: undefined,
      before: undefined,
      where: undefined
    });
  });

  it('should prevPage', () => {
    const wrapper = shallow(<Toolbar {...props}>
      <div></div>
    </Toolbar>);
    wrapper.instance().prevPage();
    expect(updateQuery.mock.calls[0][0]).toEqual(['posts']);
    expect(updateQuery.mock.calls[0][1]).toEqual({
      orderBy: undefined,
      first: undefined,
      after: undefined,
      before: "0",
      last: 10,
      where: undefined
    });
  });

  // not suport for now
  // it('should changeSize', () => {
  //   const wrapper = shallow(<Toolbar {...props}>
  //     <div></div>
  //   </Toolbar>);
  //   wrapper.instance().changeSize(20);
  //   expect(updateQuery.mock.calls[0][0]).toEqual(['posts']);
  //   expect(updateQuery.mock.calls[0][1]).toEqual({
  //     orderBy: undefined,
  //     first: undefined,
  //     after: undefined,
  //     before: "0",
  //     last: 20,
  //     where: undefined
  //   });
  // });
});