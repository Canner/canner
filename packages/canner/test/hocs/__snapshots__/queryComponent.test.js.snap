// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toolbar method should render 1`] = `
<ToolbarLayout
  Actions={<div />}
  Filter={
    <FilterGroup
      async={true}
      changeFilter={[Function]}
      deleteFilter={[Function]}
      displayedFilters={Array []}
      items={
        Object {
          "like": Object {
            "type": "number",
          },
          "share": Object {
            "type": "number",
          },
          "title": Object {
            "type": "string",
          },
        }
      }
      where={Object {}}
    />
  }
  Pagination={
    <PaginationPlugin
      async={true}
      changePage={[Function]}
      changeSize={[Function]}
      current={1}
      hasNextPage={true}
      hasPreviousPage={true}
      nextPage={[Function]}
      prevPage={[Function]}
      total={0}
    />
  }
  Sort={null}
>
  <div
    rootValue={
      Object {
        "posts": Array [
          Object {
            "id": "0",
            "title": "posts1",
          },
        ],
      }
    }
    showPagination={false}
    value={
      Array [
        Object {
          "id": "0",
          "title": "posts1",
        },
      ]
    }
  />
</ToolbarLayout>
`;
